pyomo/contrib/solver/solvers/highs.py

```diff
@@ -0,0 +1,705 @@
+        if self._last_results_object is not None:
+            self._last_results_object.solution_loader.invalidate()
+        n = len(self._pyomo_var_to_solver_var_map)
+        indices = np.arange(n)
+        costs = np.zeros(n, dtype=np.double)
+        self._objective_helpers = []
+        if obj is None:
+            sense = highspy.ObjSense.kMinimize
+            self._solver_model.changeObjectiveOffset(0)
+        else:
+            if obj.sense == minimize:
+                sense = highspy.ObjSense.kMinimize
+            elif obj.sense == maximize:
+                sense = highspy.ObjSense.kMaximize
+            else:
+                raise ValueError(f'Objective sense is not recognized: {obj.sense}')
+
+            repn = generate_standard_repn(
+                obj.expr, quadratic=False, compute_values=False
+            )
+            if repn.nonlinear_expr is not None:
+                raise IncompatibleModelError(
+                    f'Highs interface does not support expressions of degree {repn.polynomial_degree()}'
+                )
+
+            for coef, v in zip(repn.linear_coefs, repn.linear_vars):
+                v_id = id(v)
+                v_ndx = self._pyomo_var_to_solver_var_map[v_id]
+                costs[v_ndx] = value(coef)
+                if not is_constant(coef):
+                    mutable_objective_coef = _MutableObjectiveCoefficient(
+                        pyomo_var_id=v_id,
+                        var_map=self._pyomo_var_to_solver_var_map,
+                        expr=coef,
+                        highs=self._solver_model,
+                    )
+                    self._objective_helpers.append(mutable_objective_coef)
+
+            self._solver_model.changeObjectiveOffset(value(repn.constant))
+            if not is_constant(repn.constant):
+                mutable_objective_offset = _MutableObjectiveOffset(
+                    expr=repn.constant, highs=self._solver_model
+                )
+                self._objective_helpers.append(mutable_objective_offset)
+
+        self._solver_model.changeObjectiveSense(sense)
+        self._solver_model.changeColsCost(n, indices, costs)
+
+    def _postsolve(self):
+        config = self._active_config
+        timer = config.timer
+        timer.start('load solution')
+
+        highs = self._solver_model
+        status = highs.getModelStatus()
+
+        results = Results()
+        results.solution_loader = PersistentSolutionLoader(self)
+        results.timing_info.highs_time = highs.getRunTime()
+
+        self._sol = highs.getSolution()
+        has_feasible_solution = self._sol.value_valid
+        if status == highspy.HighsModelStatus.kOptimal:
+            results.solution_status = SolutionStatus.optimal
+        elif has_feasible_solution:
+            results.solution_status = SolutionStatus.feasible
+        else:
+            results.solution_status = SolutionStatus.noSolution
+
+        if status == highspy.HighsModelStatus.kNotset:
```

Comment:
Note for myself for the future - @michaelbynum made a really great `_tc_map` in `gurobi_direct` that I love. All of these redonkulous "if ... elif... elif..." statements should probably be replaced by maps.

Reply:
Wasn't me, but I agree!

Reply:
Agreed. I am not incorporating it into this one.